ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/api/impl/InMemoryLedgerReaderWriter.scala

```diff
@@ -0,0 +1,142 @@
+
+  private val currentState = new InMemoryState()
+
+  private val StartOffset: Int = 0
+
+  override def commit(correlationId: String, envelope: Array[Byte]): Future[SubmissionResult] =
+    Future {
+      val submission = Envelope
+        .openSubmission(envelope)
+        .getOrElse(throw new IllegalArgumentException("Not a valid submission in envelope"))
+      currentState.synchronized {
+        val stateInputs: Map[DamlStateKey, Option[DamlStateValue]] =
+          submission.getInputDamlStateList.asScala
+            .map(key => key -> currentState.state.get(key.toByteString))(breakOut)
+        val entryId = allocateEntryId()
+        val (logEntry, damlStateUpdates) =
+          KeyValueCommitting.processSubmission(
+            engine,
+            entryId,
+            currentRecordTime(),
+            LedgerReader.DefaultTimeModel,
+            submission,
+            participantId,
+            stateInputs
+          )
+        verifyStateUpdatesAgainstPreDeclaredOutputs(damlStateUpdates, entryId, submission)
+        val stateUpdates = damlStateUpdates.map {
+          case (damlStateKey, value) => damlStateKey.toByteString -> value
+        }
+        currentState.log += LogEntry(entryId, Envelope.enclose(logEntry).toByteArray)
+        currentState.state ++= stateUpdates
+        dispatcher.signalNewHead(currentState.log.size)
+      }
+      SubmissionResult.Acknowledged
+    }
+
+  private def verifyStateUpdatesAgainstPreDeclaredOutputs(
+      actualStateUpdates: Map[DamlStateKey, DamlStateValue],
+      entryId: DamlLogEntryId,
+      submission: DamlSubmission): Unit = {
+    val expectedStateUpdates = KeyValueCommitting.submissionOutputs(entryId, submission)
+    if (!(actualStateUpdates.keySet subsetOf expectedStateUpdates)) {
+      val unaccountedKeys = actualStateUpdates.keySet diff expectedStateUpdates
+      sys.error(
+        s"State updates not a subset of expected updates! Keys [$unaccountedKeys] are unaccounted for!")
+    }
+  }
+
+  override def events(offset: Option[Offset]): Source[LedgerRecord, NotUsed] =
+    dispatcher
+      .startingAt(
+        offset
+          .map(_.components.head.toInt)
+          .getOrElse(StartOffset),
+        OneAfterAnother[Int, List[LedgerRecord]](
+          (index: Int, _) => index + 1,
+          (index: Int) => Future.successful(List(retrieveLogEntry(index)))
+        )
+      )
+      .mapConcat { case (_, updates) => updates }
+
+  override def retrieveLedgerId(): LedgerId = ledgerId
+
+  override def currentHealth(): HealthStatus = Healthy
+
+  override def close(): Unit = ()
+
+  private val dispatcher: Dispatcher[Int] =
+    Dispatcher("in-memory-key-value-participant-state", zeroIndex = 0, headAtInitialization = 0)
+
+  private val randomNumberGenerator = new Random()
+
+  private val NamespaceLogEntries = ByteString.copyFromUtf8("L")
+
+  private def allocateEntryId(): DamlLogEntryId = {
+    val nonce: Array[Byte] = Array.ofDim(8)
```

Comment:
We should probably undo this mistake of mine and use a better entry id allocation schema for the in-memory ledgers. e.g. use monotonically increasing integer. Medium term we'll do away with the entry id though (removal of relative ids), so this isn't a big deal right now.

Reply:
Good point -- will address that in a separate PR.